In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [8]:
df = pd.read_csv(r"C:\Users\admin\PORKODI CLG\Data-Science-Learning\Task 19 - Feature Engineering\loan_data.csv")
df.head()

,age,income,loan_amount,employment_type,application_date,remarks,Loan_Default
0,25,35000,100000,Salaried,2025-01-15,Need loan for education,0
1,40,70000,200000,Business,2025-02-10,Business expansion loan,0
2,30,45000,150000,Salaried,2025-03-05,Personal expenses,1
3,50,90000,250000,Self-Employed,2025-01-20,Medical emergency,0
4,35,50000,180000,Business,2025-04-12,Home renovation,1


In [17]:
df = df.drop("remarks", axis=1)
df.head()

,age,income,loan_amount,Loan_Default,year,month,day,employment_type_Salaried,employment_type_Self-Employed
0,25,35000,100000,0,2025,1,15,True,False
1,40,70000,200000,0,2025,2,10,False,False
2,30,45000,150000,1,2025,3,5,True,False
3,50,90000,250000,0,2025,1,20,False,True
4,35,50000,180000,1,2025,4,12,False,False


In [22]:
X = df.drop("Loan_Default", axis=1)
y = df["Loan_Default"]

print("X shape:", X.shape)
print("y shape:", y.shape)

print("\nX sample:")
print(X.head())

print("\ny sample:")
print(y.head())

X shape: (10, 8)
y shape: (10,)

X sample:
   age  income  loan_amount  year  month  day  employment_type_Salaried  \
0   25   35000       100000  2025      1   15                      True   
1   40   70000       200000  2025      2   10                     False   
2   30   45000       150000  2025      3    5                      True   
3   50   90000       250000  2025      1   20                     False   
4   35   50000       180000  2025      4   12                     False   

   employment_type_Self-Employed  
0                          False  
1                          False  
2                          False  
3                           True  
4                          False  

y sample:
0    0
1    0
2    1
3    0
4    1
Name: Loan_Default, dtype: int64


In [27]:
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "Decision Tree": DecisionTreeClassifier(),
    "Random Forest": RandomForestClassifier()
}

results = []

for name, model in models.items():
    scores = cross_val_score(model, X, y, cv=3, scoring="accuracy")
    
    results.append({
        "Model": name,
        "Mean Accuracy": scores.mean(),
        "Std Dev": scores.std()
    })

results_df = pd.DataFrame(results)
print(results_df)

                 Model  Mean Accuracy   Std Dev
0  Logistic Regression       0.305556  0.039284
1        Decision Tree       0.500000  0.136083
2        Random Forest       0.472222  0.196419


In [31]:
print(results_df)

best_row = results_df.loc[results_df["Mean Accuracy"].idxmax()]
print("Best Model:")
print(best_row)

best_model = RandomForestClassifier()
best_model.fit(X, y)

sample = X.iloc[0:1]
prediction = best_model.predict(sample)

print("Prediction:", prediction[0])

                 Model  Mean Accuracy   Std Dev
0  Logistic Regression       0.305556  0.039284
1        Decision Tree       0.500000  0.136083
2        Random Forest       0.472222  0.196419
Best Model:
Model            Decision Tree
Mean Accuracy              0.5
Std Dev               0.136083
Name: 1, dtype: object
Prediction: 0


## Task 23 - Model Comparison Report

### Objective
Compare 3 machine learning models using cross-validation and select the best model.

---

### Dataset
Loan default dataset with:
- age, income, loan_amount
- employment_type
- year, month, day  
Target: Loan_Default

---

### Preprocessing
- Converted employment_type into numbers
- Removed remarks column
- Split data into X and y

---

### Models Used
- Logistic Regression
- Decision Tree
- Random Forest

---

### Evaluation
- Cross-validation (cv = 3 )
- Metric: Accuracy

---

### Results

                 Model  Mean Accuracy   Std Dev
0  Logistic Regression       0.305556  0.039284
1        Decision Tree       0.500000  0.136083
2        Random Forest       0.472222  0.196419

---

### Best Model
- Best Model: Random Forest Classifier based on highest mean accuracy.
---

### Conclusion
Best model is selected based on highest accuracy and stable results.